core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java

```diff
@@ -229,4 +252,93 @@ public void getSubRef() throws IOException {
         assertTrue(refs.tail.head.isGraph());
     }
 
+    private final Token smallSeq = seq(any("b"), any("c"));
+
+    @Test
+    public void getAllRootsSingle() throws IOException {
+        final Token topSeq = seq(any("a"), smallSeq);
+        final ParseResult result = topSeq.parse(stream(1, 2, 3), enc());
+        assertTrue(result.succeeded);
+        final ParseItemList seqs = ByToken.getAllRoots(result.environment.order, smallSeq);
+        assertEquals(1, seqs.size);
+        assertEquals(smallSeq, seqs.head.getDefinition());
+        final ParseValue c = seqs.head.asGraph().head.asValue();
+        assertEquals(3, c.asNumeric().intValue());
+        assertEquals(2, c.getOffset());
+    }
+
+    @Test
+    public void getAllRootsMulti() throws IOException {
+        final Token topSeq = seq(any("a"), smallSeq, smallSeq);
+        final ParseResult result = topSeq.parse(stream(1, 2, 3, 2, 3), enc());
+        assertTrue(result.succeeded);
+        final ParseItemList seqs = ByToken.getAllRoots(result.environment.order, smallSeq);
+        assertEquals(2, seqs.size);
+        assertEquals(smallSeq, seqs.head.getDefinition());
+        assertEquals(smallSeq, seqs.tail.head.getDefinition());
+        final ParseValue c1 = seqs.head.asGraph().head.asValue();
+        assertEquals(3, c1.asNumeric().intValue());
+        final ParseValue c2 = seqs.tail.head.asGraph().head.asValue();
+        assertEquals(3, c2.asNumeric().intValue());
+        assertNotEquals(seqs.head.asGraph().head, seqs.tail.head.asGraph().head);
+    }
+
+    private HashSet<ParseItem> makeSet(ParseItemList seqs) {
+        final HashSet<ParseItem> items = new HashSet<>();
+        for (ParseItemList current = seqs; current != null && !current.isEmpty(); current = current.tail) {
+            items.add(current.head);
```

Comment:
I don't think that any of our `ParseItem`s implement a `hashCode()`, are you sure this actually works as expected?